Flink 33386 review debug 0124

RocMarshal opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*


## Brief change log

*(for example:)*
  - *The TaskInfo is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *TaskManagers retrieve the TaskInfo from the blob cache*


## Verifying this change

Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluster with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

RocMarshal committed:
[Hotfix][runtime] Enrich the toString method of SlotSharingGroup class

RocMarshal committed:
[FLINK-33386][runtime] Support tasks balancing at slot level for Default Scheduler

RocMarshal committed:
[FLINK-33874][runtime] Support resource request wait mechanism at DefaultDeclarativeSlotPool side for Default Scheduler

RocMarshal committed:
[FLINK-33875][runtime] Support slots wait mechanism at DeclarativeSlotPoolBridge side for Default Scheduler

RocMarshal committed:
[Hotfix][runtime] Add '@Internal' annotation for TaskManagerLoadBalanceMode#loadFromConfiguration

RocMarshal committed:
[FLINK-33387][runtime] Introduce the abstraction and the interface about loading

RocMarshal committed:
[FLINK-33388][runtime] Support tasks balancing at TaskExecutor level for Default Scheduler

RocMarshal commented:
CC @1996fanrui updated.

1996fanrui commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/AbstractSlotSharingStrategyTest.java:
```diff
@@ -0,0 +1,181 @@
+import org.apache.flink.runtime.scheduler.strategy.TestingSchedulingTopology;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.flink.shaded.guava32.com.google.common.collect.Lists;
+import org.apache.flink.shaded.guava32.com.google.common.collect.Sets;
```
这两行应该在第二个 commit 里，但是第 3 个 commit 里进行了更改。

RocMarshal replied:
done

1996fanrui commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerBuilder.java:
```diff
@@ -102,7 +102,9 @@ public AdaptiveSchedulerBuilder(
                         DEFAULT_TIMEOUT,
-                        rpcTimeout);
+                        rpcTimeout,
+                        null,
```
这里不能为 null，构造器里有 非空检查。目前所有相关的 test 都会报错。

RocMarshal replied:
done

1996fanrui commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java:
```diff
@@ -523,7 +524,9 @@ void testNumRestartsMetric() throws Exception {
                         Time.minutes(10),
-                        Time.minutes(10));
+                        Time.minutes(10),
+                        null,
```
这里不能为 null，构造器里有 非空检查。目前所有相关的 test 都会报错。

RocMarshal replied:
done

1996fanrui commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java:
```diff
@@ -2111,7 +2114,9 @@ private static DefaultDeclarativeSlotPool createDeclarativeSlotPool(
                 Time.fromDuration(idleSlotTimeout),
-                Time.fromDuration(DEFAULT_TIMEOUT));
+                Time.fromDuration(DEFAULT_TIMEOUT),
+                null,
```
这里不能为 null，构造器里有 非空检查。目前所有相关的 test 都会报错。

RocMarshal replied:
done

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPoolServiceFactory.java:
```diff
@@ -40,8 +43,18 @@ public DeclarativeSlotPoolServiceFactory(Clock clock, Time idleSlotTimeout, Time
+                jobId,
+                declarativeSlotPoolFactory,
+                clock,
+                idleSlotTimeout,
+                rpcTimeout,
+                slotRequestMaxInterval,
+                componentMainThreadExecutor);
```
`idleSlotTimeout` 与 `slotRequestMaxInterval` 以及 `slotBatchAllocatable` 这类参数都类似吧？

感觉他们都应该通过 构造器传参，而不是 `createSlotPoolService` 这个接口。

ComponentMainThreadExecutor 放到接口的参数可能是合理的。

RocMarshal replied:
done

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPoolBridge.java:
```diff
@@ -200,26 +223,43 @@ protected void onReleaseTaskManager(ResourceCounter previouslyFulfilledRequireme
+                        receivedSlots,
+                        pendingRequests.values(),
+                        getDeclarativeSlotPool().getTaskExecutorsLoadingWeight());
+        if (requestSlotMatches.size() >= pendingRequests.size()) {
+            Preconditions.checkState(
+                    requestSlotMatches.size() == pendingRequests.size(),
+                    "The number of matched slots is not equals to the pendingRequests.");
```
这个 if 和 checkState 的条件很奇怪。

这里其实就是三个代码分支:
- `<` : 直接 return
- `>` : 抛异常
- `==` : 走逻辑

RocMarshal replied:
done

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:
```diff
@@ -127,7 +145,25 @@ public void increaseResourceRequirementsBy(ResourceCounter increment) {
         }
         totalResourceRequirements = totalResourceRequirements.add(increment);
```
注：这里 看起来有一些代码没改动，其实第三个 commit 里改了。第四个 commit 里撤回了。需要处理下。

RocMarshal replied:
done

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:
```diff
@@ -127,7 +145,25 @@ public void increaseResourceRequirementsBy(ResourceCounter increment) {
+        doDeclareResourceRequirements();
+    }
+
+    private void doDeclareResourceRequirements() {
```
这个方法放到 `decreaseResourceRequirementsBy` 之后吧，公共方法放到 `increase` 和 `decrease` 之间很奇怪。

RocMarshal replied:
done.

RocMarshal commented:
checking ci for every independent commit:   
https://github.com/apache/flink/pull/24229

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/slots/ResourceRequirement.java:
```diff
@@ -32,12 +36,26 @@ public class ResourceRequirement implements Serializable {
+        Preconditions.checkArgument(numberOfRequiredSlots >= 0);
+        Preconditions.checkArgument(
+                numberOfRequiredSlots == loadingWeights.size(),
+                "The number of loading weight info must be equals to numberOfRequiredSlots.");
```
把两个具体的 size 都打出来吧

RocMarshal replied:
使用新的LoadableResourceProfile后，这里就不在需要，移出了。

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:
```diff
@@ -540,7 +619,10 @@ private void releaseSlots(Iterable<AllocatedSlot> slotsToReturnToOwner, Throwabl
             final ResourceProfile matchingResourceProfile =
```
// TODO 没 match loading 就去 减资源了，有可能 match 不上。

RocMarshal replied:
updated.

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/JobScopedResourceTracker.java:
```diff
@@ -91,19 +101,21 @@ private Optional<ResourceProfile> findMatchingRequirement(ResourceProfile resour
                 resourceToRequirementMapping::getNumFulfillingResources);
```
这里 match 需要加 loading

RocMarshal replied:
updated.

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/JobScopedResourceTracker.java:
```diff
@@ -91,19 +101,21 @@ private Optional<ResourceProfile> findMatchingRequirement(ResourceProfile resour
-    public void notifyLostResource(ResourceProfile resourceProfile) {
+    public void notifyLostResource(ResourceProfile resourceProfile, LoadingWeight loadingWeight) {
         Preconditions.checkNotNull(resourceProfile);
         if (excessResources.getResourceCount(resourceProfile) > 0) {
```
这个 get 时，也需要看 loading。

防止 其他 loading=3 的 slot 有资源，却去 减一个 loading=4 的 资源

RocMarshal replied:
updated.

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/ResourceCounter.java:
```diff
@@ -18,28 +18,47 @@
+    public static final Logger LOG = LoggerFactory.getLogger(ResourceCounter.class);
 
     private final Map<ResourceProfile, Integer> resources;
 
-    private ResourceCounter(Map<ResourceProfile, Integer> resources) {
+    private final Map<ResourceProfile, List<LoadingWeight>> loadingWeightsMap;
+
```
可能需要用 ResourceProfile + LoadingWeight 做为 key。

RocMarshal replied:
done

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/slots/ResourceRequirement.java:
```diff
@@ -32,12 +36,26 @@ public class ResourceRequirement implements Serializable {
+        this.loadingWeights =
+                new ArrayList<LoadingWeight>() {
+                    {
+                        addAll(loadingWeights);
```
不能使用 new ArrayList<>(loadingWeights) 吗？

我本地测试没问题：

<img width="751" alt="image" src="https://github.com/RocMarshal/flink/assets/38427477/ae7a7a91-fc73-4217-9cc2-f293e8d3e60b">

RocMarshal replied:
已经移除

hackergin committed:
[hotfix][doc] Update the Kafka DDL example in the hive catalog page to the new parameter style

affo committed:
[FLINK-20398][e2e] Migrate test_batch_sql.sh to Java e2e tests framework

1996fanrui commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java:
```diff
@@ -1085,6 +1095,146 @@ void testSlotRequestMaxIntervalAndComponentMainThreadExecutorPassing() throws Ex
+                // Arguments.of(SchedulerType.AdaptiveBatch, JobType.STREAMING,
+                // TaskManagerLoadBalanceMode.SLOTS, true, false),
+                // Arguments.of(SchedulerType.AdaptiveBatch, JobType.STREAMING,
+                // TaskManagerLoadBalanceMode.TASKS, true, true),
```
These lines are not needed. From the naming, most of developers know `SchedulerType.AdaptiveBatch` doesn't support `JobType.STREAMING`.

RocMarshal replied:
updated.

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/loading/LoadingWeight.java:
```diff
@@ -0,0 +1,66 @@
+
+    LoadingWeight EMPTY = new DefaultLoadingWeight(0f);
+
+    static LoadingWeight ofDefaultLoadingWeight(float loading) {
```
It's better to that interface doesn't care about the specific implementation. 

All callers can new `DefaultLoadingWeight` directly instead of calling the interface.

RocMarshal replied:
updated.

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/loading/LoadingWeight.java:
```diff
@@ -0,0 +1,66 @@
+     * @return a new multiplied object.
+     */
+    default LoadingWeight multiply(float factor) {
+        return LoadingWeight.ofDefaultLoadingWeight(factor * getLoading());
```
multiply should be an interface, and all implementations can overwrite it.

For example: `DefaultLoadingWeight` will new a new `DefaultLoadingWeight`. We may define new implementation, it can return the type of itself.

RocMarshal replied:
updated.

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/loading/DefaultLoadingWeight.java:
```diff
@@ -0,0 +1,71 @@
+
+    @Override
+    public String toString() {
+        return "DefaultLoadingWeight{" + "loading=" + loading + '}';
```
The first `" + "` is not necessary.

RocMarshal replied:
updated.

affo committed:
[FLINK-20398][test] solve deprecation in BatchSQLTest source table creation

drymatini committed:
[FLINK-35353][docs-zh]Translate "Profiler" page into Chinese

This closes #24822.

XComp committed:
[FLINK-35550][runtime] Adds new component RescaleManager that handles the rescaling logic to improve code testability and extensibility

Rescaling is a state-specific functionality. Moving all the logic into Executing state allows
us to align the resource controlling in Executing state and WaitingForResources state in a future effort.

XComp committed:
[FLINK-35551][runtime] Adds RescaleManager#onTrigger

- integrates trigger logic in DefaultRescaleManager
- a new parameter jobmanager.adaptive-scheduler.max-delay-for-scale-trigger

wForget committed:
Fix artifactId of flink-migration-test-utils module

zoltar9264 committed:
[FLINK-32086][checkpointing] Normalize file-merging sub dir

zoltar9264 committed:
[FLINK-32086][checkpointing] Cleanup orphan file-merging managed directory on exit of TM

This closes #24933

zoltar9264 committed:
[FLINK-32086][checkpointing] Fix deletion fail of file-merging managed dir

RocMarshal committed:
[hotfix][runtime] Change the location for test cases files of SlotSelectionStrategy

ZmmBigdata committed:
[hotfix][docs] Fix the Incorrect syntax `GROUPING SET` in `group-agg.md`(#24947)

dawidwys committed:
[FLINK-35619] Window rank query fails with 'must call validate first' (#24940)

hackergin committed:
[FLINK-33936][table] Outputting Identical Results in Mini-Batch Aggregation with Set TTL

hackergin committed:
[hotfix][table] Fix warning message format in WorkflowSchedulerFactoryUtil

SinBex committed:
[FLINK-35656][hive] Fix the issue that Hive Source incorrectly set max parallelism in dynamic inference mode

This closes #24962.

TanYuxin-tyx committed:
[FLINK-35658][runtime] Fix the issue that the external tier of hybrid shuffle can not work with an UnknownInputChannel

lsyldliu committed:
[FLINK-35659][testutils] Fix TestFileSystem Connector error setting execution mode

RocMarshal committed:
[hotfix][streaming] Move the TestProcessingTimeService.java down to the corresponding test package.

lsyldliu committed:
[FLINK-35659][table] Remove unnecessary parameter customScheduleTime for embedded scheduler

TanYuxin-tyx committed:
Revert "[FLINK-33668][runtime] Avoid the potential deadlock during reserving segments"

This reverts commit 4171b980717e5dc313d781210a5bbfd35449ee50.

TanYuxin-tyx committed:
Revert "[FLINK-33668][runtime] Close the producer client in case of the result partition is not finished properly"

This reverts commit c1ea971129aa1f91c07cbb250ef4e377a7d67dec.

TanYuxin-tyx committed:
Revert "[FLINK-33668][runtime] Make the SortBuffer be able to return unused memory segments"

This reverts commit ac2ec3496400b594ffe5c711bc1b6343999a8479.

TanYuxin-tyx committed:
Revert "[FLINK-33668][runtime] Decouple the network memory and job topology on result partition side"

This reverts commit 88138d08e731d3084d59bf14cc2fb51bdf4afbd8.

TanYuxin-tyx committed:
Revert "[FLINK-33668][runtime] Decouple the network memory and job topology on input gate side"

This reverts commit 2e257896661ad5e0dca521fa274cad3a0bae7a61.

hackergin committed:
[FLINK-35479][e2e] Add end-to-end test for materialized table

TanYuxin-tyx committed:
Revert "[FLINK-33668][runtime] Redefine the redistribution logic of the NetworkBufferPool"

This reverts commit 012b8936e119d1bfadadee3182cf86d25c6dd019.

showuon committed:
[hotfix][docs] add missing output method in Context interface

lsyldliu committed:
[FLINK-35646][docs] Fix the error description of token parameter

lsyldliu committed:
[FLINK-35646][table] Exclude embedded workflow scheduler related rest api from docs

lsyldliu committed:
[FLINK-35646][docs] Add docs for materialized table refresh api

TanYuxin-tyx committed:
[hotfix][runtime] Fix the issue of registering and unregistering a job for shuffle master

reswqa committed:
[FLINK-35673][test] Correct the SELF type of TestFileSourceBuilder

reswqa committed:
[FLINK-35682][release] Update version to 2.0-SNAPSHOT

reswqa committed:
[FLINK-35682][release] Add new version 2.0 and update GHA nightly build versions

jnh5y committed:
[FLINK-35585] Add documentation for DISTRIBUTED BY (#24929)

reswqa committed:
[FLINK-35682][docs] Include `release-1.20` for building documentation

lsyldliu committed:
[FLINK-35692][docs] Remove hugo_extended_0.110.0_Linux-64bit.tar.gz file introduced by wrong operation

lsyldliu committed:
[FLINK-35692][docs] Fix the default version of sql gateway OpenAPI

lsyldliu committed:
Revert "[FLINK-35189][connectors/filesystem] Modify the visibility of filesystem connector related methods to protected"

This reverts commit e3cda01cac737c97a104feae0e35da1eb91a8751.

lsyldliu committed:
[FLINK-35691][testutils] Optimize test-filesystem logic to support streaming and batch read

lsyldliu committed:
[FLINK-35691][table] Fix partition.fields.#.date-formatter option verify logic

lsyldliu committed:
[FLINK-35691][table] Fix unexpected behavior of repeated suspend and resume materialized table

ferenc-csaky committed:
[FLINK-35699][k8s] Fix shading Jackson through fabric8

reswqa committed:
[hotfix] Update OpenAPI spec version

lsyldliu committed:
[FLINK-35691][table] Fix drop table statement can drop materialized table

lsyldliu committed:
[FLINK-35691][table] Fix cluster info displayed incorrectly on SqlClient when RowFormat is PLAIN_TEXT

dawidwys committed:
[FLINK-35687] JSON_QUERY should return a well formatted nested objects/arrays for ARRAY<STRING> (#24976)

xuyangzhong committed:
[FLINK-35271][docs] Add doc for DESCRIBE JOB <JOB_ID> statement

This closes #24984

Stephen0421 committed:
[FLINK-35702][parquet] support parquet read nested in nested (#24991)

empathy87 committed:
[hotfix][parquet] Fix row group filter support issue (#24979)

hackergin committed:
[FLINK-35734][table] Fix periodic refresh job naming

hackergin committed:
[FLINK-35734][table] Do not override the user-defined checkpoint interval in continuous mode

xuyangzhong committed:
[FLINK-35498][table] Fix unexpected argument name conflicts when extracting method parameter names from UDF

This closes #24890

Stephen0421 committed:
[hotfix][parquet] Fix dependences and variable name problems in FLINK-35702 (#25004)
